spec.bs

```diff
@@ -343,6 +344,83 @@ Each {{AppHistory}} object has an associated <dfn for="AppHistory">current index
 
 <h3 id="ongoing-state">Ongoing navigation tracking</h3>
 
+<xmp class="idl">
+[Exposed=Window]
+interface AppHistoryTransition {
+  readonly attribute AppHistoryNavigationType navigationType;
+  readonly attribute AppHistoryEntry from;
+  readonly attribute Promise<undefined> finished;
+
+  Promise<undefined> rollback(optional AppHistoryNavigationOptions options = {});
+};
+</xmp>
+
+<dl class="domintro">
+  <dt><code>{{Window/appHistory}}.{{AppHistory/transition}}</code>
+  <dd>
+    <p>An {{AppHistoryTransition}} object representing any ongoing navigation that hasn't yet reached the {{AppHistory/navigatesuccess}} or {{AppHistory/navigateerror}} stage, if one exists, or null if there is no such transition ongoing.
+
+    <p>Since {{AppHistory/current|appHistory.current}} (and other properties like {{Location/href|location.href}}) are updated immediately upon navigation, this {{AppHistory/transition|appHistory.transition}} property is useful for determining when such navigations are not yet fully settled, according to any promises passed to {{AppHistoryNavigateEvent/respondWith()|event.respondWith()}}.
+  </dd>
+
+  <dt><code>{{Window/appHistory}}.{{AppHistory/transition}}.{{AppHistoryTransition/navigationType}}</code></dt>
+  <dd>
+    <p>One of "{{AppHistoryNavigationType/reload}}", "{{AppHistoryNavigationType/push}}", "{{AppHistoryNavigationType/replace}}", or "{{AppHistoryNavigationType/traverse}}", indicating what type of navigation this transition is for.
+  </dd>
+
+  <dt><code>{{Window/appHistory}}.{{AppHistory/transition}}.{{AppHistoryTransition/from}}</code></dt>
+  <dd>
+    <p>The {{AppHistoryEntry}} from which the transition is coming. This can be useful to compare against {{AppHistory/current|appHistory.current}}.
+  </dd>
+
+  <dt><code>{{Window/appHistory}}.{{AppHistory/transition}}.{{AppHistoryTransition/finished}}</code></dt>
```

Comment:
The main point is to give a promise-based alternative to `navigatesuccess` and `navigateerror`, since promises can be more convenient than events if you only want to subscribe once.

https://github.com/WICG/app-history/issues/11 also has some discussion.